src/core/services/compiler/compiler.js

```diff
@@ -81,249 +87,384 @@ angular
+   * for more information.
+   *
+   * If disabled (false), the compiler assigns the value of each of the bindings to the
+   * properties of the controller object before the constructor of this object is called.
+   *
+   * If enabled (true) the behavior depends on the AngularJS version used:
+   *
+   *     - `<1.5.10` - bindings are pre-assigned
+   *     - `>=1.5.10 <1.7` - behaves like set to whatever `$compileProvider.preAssignBindingsEnabled()` reports; if
+   *                         the `$compileProvider` flag wasn't set manually, it defaults to pre-assigning bindings
+   *                         with AngularJS `1.5.x` & to calling the constructor first with AngularJS `1.6.x`.
+   *     - `>=1.7` - the compiler calls the constructor first before assigning bindings
+   *
+   * The default value is `false` but will change to `true` in AngularJS Material 1.2.
+   *
+   * It is recommended to set this flag to `true` in AngularJS Material 1.1.x; the only reason it's not set that way
+   * by default is backwards compatibility. Not setting the flag to `true` when
+   * `$compileProvider.preAssignBindingsEnabled()` is set to `false` (i.e. default behavior in AngularJS 1.6.0 or newer)
+   * makes it hard to unit test Material Dialog/Toast controllers using the `$controller` helper as it always follows
+   * the `$compileProvider.preAssignBindingsEnabled()` value.
+   *
+   * @deprecated
+   * sinceVersion="1.2.0"
+   * removeVersion="1.3.0"
+   *
+   * This method and the option to assign the bindings before calling the controller's constructor
+   * will be removed in v1.3.0.
+   */
+  var respectPreAssignBindingsEnabled = false;
+  this.respectPreAssignBindingsEnabled = function(enabled) {
+    if (angular.isDefined(enabled)) {
+      respectPreAssignBindingsEnabled = enabled;
+      return this;
+    }
 
-  /** @private @const {!angular.$templateRequest} */
-  this.$templateRequest = $templateRequest;
+    return respectPreAssignBindingsEnabled;
+  };
 
-  /** @private @const {!angular.$injector} */
-  this.$injector = $injector;
+  /**
+   * @ngdoc method
+   * @name  $mdCompilerProvider#preAssignBindingsEnabled
+   *
+   * @returns {*} current preAssignBindingsEnabled state
+   *
+   * @kind function
+   *
+   * @description
+   * This method returns `true` if Material-specific (dialogs/toasts) controllers have bindings pre-assigned in
+   * controller constructors and `false` otherwise.
+   *
+   * Note that this doesn't affect directives/components created via regular AngularJS methods which constitute most
+   * Material & user-created components; their behavior can be checked via `$compileProvider.preAssignBindingsEnabled()`
+   * in AngularJS `>=1.5.10 <1.7.0`.
+   *
+   * @deprecated
+   * sinceVersion="1.2.0"
+   * removeVersion="1.3.0"
+   *
+   * This method will be removed in v1.3.0.
+   */
+  this.preAssignBindingsEnabled = function() {
```

Comment:
Also, maybe change the name to indicate it's a getter? Like `getPreAssignBindingsEnabled`.